shared/src/pool_cache.rs

```diff
@@ -0,0 +1,461 @@
+            }
+        };
+
+        let mut cache_hits = Vec::new();
+        let mut cache_misses = HashSet::new();
+        let last_update_block;
+        {
+            let mut mutexed = self.mutexed.lock().unwrap();
+            for &pair in &pairs {
+                match mutexed.get(pair, block) {
+                    Some(pools) => cache_hits.extend_from_slice(&pools),
+                    None => {
+                        cache_misses.insert(pair);
+                    }
+                }
+            }
+            last_update_block = mutexed.last_update_block;
+        }
+
+        tracing::debug!(
+            "{} cache misses out of {} total",
+            cache_misses.len(),
+            pairs.len()
+        );
+
+        if cache_misses.is_empty() {
+            return Ok(cache_hits);
+        }
+
+        let cache_miss_block = block.unwrap_or(last_update_block);
+        let uncached_pools = self
+            .inner
+            .fetch(
+                cache_misses.clone(),
+                BlockNumber::Number(cache_miss_block.into()),
+            )
+            .await?;
+
+        {
+            let mut mutexed = self.mutexed.lock().unwrap();
+            mutexed.insert(cache_miss_block, pairs.iter().copied(), &uncached_pools);
+        }
+
+        cache_hits.extend_from_slice(&uncached_pools);
+        Ok(cache_hits)
+    }
+}
+
+#[async_trait::async_trait]
+impl Maintaining for PoolCache {
+    async fn run_maintenance(&self) -> Result<()> {
+        let block = current_block::block_number(&self.block_stream.borrow())?;
+        self.update_cache(block)
+            .await
+            .context("failed to update pool cache")
+    }
+}
+
+#[derive(Debug)]
+struct Mutexed {
+    recently_used: LruCache<TokenPair, ()>,
+    // For quickly finding at which block a pair is cached.
+    cached_most_recently_at_block: HashMap<TokenPair, u64>,
+    // Tuple ordering allows us to efficiently construct range queries by block.
+    pools: BTreeMap<(u64, TokenPair), Vec<Pool>>,
+    // The last block at which the automatic cache updating happened.
+    last_update_block: u64,
+}
+
+impl Mutexed {
+    fn new(pairs_lru_size: usize, current_block: u64) -> Mutexed {
+        Self {
+            recently_used: LruCache::new(pairs_lru_size),
+            cached_most_recently_at_block: HashMap::new(),
+            pools: BTreeMap::new(),
+            last_update_block: current_block,
+        }
+    }
+}
+
+impl Mutexed {
+    fn get(&mut self, pair: TokenPair, block: Option<u64>) -> Option<&[Pool]> {
+        self.recently_used.put(pair, ());
+        let block = block.or_else(|| self.cached_most_recently_at_block.get(&pair).copied())?;
+        self.pools.get(&(block, pair)).map(Vec::as_slice)
+    }
+
+    fn insert(&mut self, block: u64, pairs: impl Iterator<Item = TokenPair>, pools: &[Pool]) {
+        for pair in pairs {
+            self.cached_most_recently_at_block.insert(pair, block);
```

Comment:
What do we mean by "recently" here? It looks like we could cache a pair for an old block (e.g. 25 blocks in the past) and would consider this most recently. Should we check that cached_most_recently_at_block is not larger than block?

Also maybe add a unit test for this case?

Reply:
Good point. I think we should have another parameter that decides the maximum age of an entry you can get when requesting "latest". This would be decoupled from the number of blocks cache entries stay alive for.

Reply:
I still think we should only insert `block` here if it's higher than what we already have saved. Otherwise, we could query for a very old block (would override `cached_most_recently_at_block`) and then would cache miss despite having information available in a "recent" block. Not a super important edge case, but feels easy enough to do this check here.